### PR TITLE
Add error recovery funcionality

### PR DESCRIPTION
**Pull request by VarS311**
PR for issue #66 

Works on two levels, Declarations and Statements.

### CODE REVIEW

File: apex-squid/src/test/java/org/fundacionjala/enforce/sonarqube/apex/parser/grammar/ApexGrammarClassOrInterfaceMemberTest.java [hunk 1]
```diff
@@ -42,4 +42,28 @@ public void positiveRules() {
                 .matches("@someAnnotation\n"
                         + "public global static void aMethodWithLoops(list<integer> collection) {}");
     }
+    
+    @Test
+    public void positiveRecoveredRules() {
```
Comment (JimmyRojas) [inline apex-squid/src/test/java/org/fundacionjala/enforce/sonarqube/apex/parser/grammar/ApexGrammarClassOrInterfaceMemberTest.java]: valid UT ?

File: apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/grammar/buildersource/Declaration.java [hunk 2]
```diff
@@ -142,12 +144,20 @@ private static void explicitConstructorInvocation(LexerfulGrammarBuilder grammar
...
      */
-    private static void propertyDeclaration(LexerfulGrammarBuilder grammarBuilder) {
-        grammarBuilder.rule(PROPERTY_DECLARATION).is(TYPE,
-                grammarBuilder.firstOf(ALLOWED_KEYWORDS_AS_IDENTIFIER, SPECIAL_KEYWORDS_AS_IDENTIFIER),
-                LBRACE,
-                ACCESSOR_DECLARATIONS,
-                RBRACE);
+    private static void propertyDeclaration(LexerfulGrammarBuilder grammarBuilder, boolean errorRecoveryEnabled) {
+        if (errorRecoveryEnabled) {
+            grammarBuilder.rule(PROPERTY_DECLARATION).is(TYPE,
```
Comment (JimmyRojas) [inline apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/grammar/buildersource/Declaration.java]: duplicated code

File: apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/grammar/buildersource/Declaration.java [hunk 3]
```diff
@@ -142,12 +144,20 @@ private static void explicitConstructorInvocation(LexerfulGrammarBuilder grammar
...
+                    grammarBuilder.firstOf(ALLOWED_KEYWORDS_AS_IDENTIFIER, SPECIAL_KEYWORDS_AS_IDENTIFIER),
+                    LBRACE,
+                    ACCESSOR_DECLARATIONS,
+                    RBRACE);
+        } else {
+            grammarBuilder.rule(PROPERTY_DECLARATION).is(TYPE,
```
Comment (JimmyRojas) [inline apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/grammar/buildersource/Declaration.java]: original code

**Comment by VarS311** [comment]
@JimmyRojas Fixed the observations and the bug I had earlier. Ready to merge.

File: apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/grammar/buildersource/Declaration.java [hunk 4]
```diff
@@ -2,7 +2,6 @@
  * Copyright (c) Fundacion Jala. All rights reserved.
```
Comment (JimmyRojas) [inline apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/grammar/buildersource/Declaration.java:2]: you were adding the functionality to

```
private static void propertyDeclaration(LexerfulGrammarBuilder grammarBuilder, boolean errorRecoveryEnabled) {
...
 if (errorRecoveryEnabled) {
 ...
 } else {
 ...
 }
...
}
```

Why did you remove it?

Comment (VarS311) [inline apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/grammar/buildersource/Declaration.java:2]: @JimmyRojas Mmm yeah I was considering adding the recovery functionality for the properties too because their braces ("{}") were causing some issues (the Block statement recovery wasn't handling them), but I removed it because in case there would be an error for a property it would be skipped and handled by the Declaration recovery, so it wasn't necessary

Comment (JimmyRojas) [inline apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/grammar/buildersource/Declaration.java:2]: ok, sounds good

**Comment by JimmyRojas** [comment]
I don't have more comments, @ziur do you have some one?

File: apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/ApexGrammar.java [hunk 5]
```diff
@@ -28,23 +29,26 @@ private ApexGrammar() {
      * It is the main method of grammar. Here all other grammars are
      * constructed.
      *
+     * @param errorRecoveryEnabled error recovery status
      * @return the grammar
      */
-    public static Grammar create() {
+    public static Grammar create(boolean errorRecoveryEnabled) {
```
Comment (ziur) [inline apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/ApexGrammar.java]: Why is optional the parsing error recovery mechanism? I think it should be always enabled because the idea is try to provide as much a possible information and don't crash with the first syntax error. The errorRecoveryEnabled flag has added an unneeded complexity in the rules section.

Comment (VarS311) [inline apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/ApexGrammar.java]: Right, I just took that part from the other plugin implementations, since more than one of them used it, but it won't be hard to change it. Should I remove it, then? @ziur @JimmyRojas

Comment (ziur) [inline apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/ApexGrammar.java]: Yes please, we should keep it as simple as possible, if at some moment we need something similar we will think about it.

**Comment by JimmyRojas** [comment]
Looks good to me, @ziur it is up to you

**Comment by ziur** [comment]
:+1: ...
